src/zenml/deployers/base_deployer.py

```diff
@@ -0,0 +1,996 @@
+        provisioning deployments using the deployer. It is used to serve
+        a pipeline snapshot as an HTTP deployment, or update an existing
+        deployment instance with the same name. The method returns a
+        DeploymentResponse object that is a representation of the
+        external deployment instance.
+
+        Args:
+            snapshot: The pipeline snapshot to serve as an HTTP deployment.
+            stack: The stack the pipeline will be served on.
+            deployment_name_or_id: Unique name or ID for the deployment.
+                This name must be unique at the project level.
+            replace: If True, it will update in-place any existing pipeline
+                deployment instance with the same name. If False, and the pipeline
+                deployment instance already exists, it will raise a
+                DeploymentAlreadyExistsError.
+            timeout: The maximum time in seconds to wait for the pipeline
+                deployment to be provisioned. If provided, will override the
+                deployer's default timeout.
+
+        Raises:
+            DeploymentAlreadyExistsError: if the deployment already
+                exists and replace is False.
+            DeploymentProvisionError: if the deployment fails.
+            DeploymentSnapshotMismatchError: if the pipeline snapshot
+                was not created for this deployer.
+            DeploymentNotFoundError: if the deployment with the
+                given ID is not found.
+            DeployerError: if an unexpected error occurs.
+
+        Returns:
+            The DeploymentResponse object representing the deployed
+            deployment.
+        """
+        client = Client()
+
+        settings = cast(
+            BaseDeployerSettings,
+            self.get_settings(snapshot),
+        )
+
+        timeout = timeout or settings.lcm_timeout
+        auth_key = settings.auth_key
+        if not auth_key and settings.generate_auth_key:
+            auth_key = self._generate_auth_key()
+
+        if snapshot.stack and snapshot.stack.id != stack.id:
+            # When a different stack is used then the one the snapshot was
+            # created for, the container image may not have the correct
+            # dependencies installed, which leads to unexpected errors during
+            # deployment. To avoid this, we raise an error here.
+            raise DeploymentSnapshotMismatchError(
+                f"The pipeline snapshot with ID '{snapshot.id}' "
+                f"was not created for the stack {stack.name} and might not "
+                "have the correct dependencies installed. This may "
+                "lead to unexpected behavior during deployment. Please switch "
+                f"to the correct active stack '{snapshot.stack.name}' or use "
+                "a different snapshot."
+            )
+
+        try:
+            # Get the existing deployment
+            deployment = client.get_deployment(
```

Comment:
In general I'd say this replace if exists logic is best implemented in the server, to reduce the possibility for race conditions. Not sure how complicated it would be to have the `replace` flag in the request/update models in your case.

Reply:
It's a good idea, but it'll take me too long to refactor the whole code using this.